### PR TITLE
Arian's Branch

### PR DESCRIPTION
**Pull request by mnoukhov**

### CODE REVIEW

File: examples/summarization/inference_compare.py [hunk 1]
```diff
@@ -187,39 +188,42 @@ def reward_model(accelerator, model, inputs):
...
-            #TODO 
-            #create input ids that are right-padded with tokenizer.pad_token_id
-            #create corresponding attention mask
-            #create labels corresponding to the generated tokens (not including prompt) 
-            #should have -100 (label_pad_token) everywhere else
-            reward_model_input = {
-                "input_ids": None,
-                "attention_mask": generated_attention_mask,
-                "labels": None,
+            generated_texts = tokenizer.batch_decode(generated_sequences, skip_special_tokens=True)
```
Comment (mnoukhov) [inline examples/summarization/inference_compare.py]: You might need to add a space between `TLDR:` and the continuation to be consistent with the supervised dataset. Can you look into this?

File: examples/summarization/inference_compare.py [hunk 2]
```diff
@@ -187,39 +188,42 @@ def reward_model(accelerator, model, inputs):
...
+            tokenizer.padding_side = "right"
+            generated_encodings = tokenizer(
+                generated_texts, padding=True, truncation=True, max_length=script_args.seq_length, return_tensors="pt"
+            )
+
+            generated_labels = generated_encodings["input_ids"].clone()
+            generated_labels[generated_labels == tokenizer.pad_token_id] = -100
+            prompt_lens = inputs["attention_mask"].sum(-1)
+            for i, prompt_len in enumerate(prompt_lens):
+                generated_labels[i * 2 : i * 2 + 2, :prompt_len] = -100
```
Comment (mnoukhov) [inline examples/summarization/inference_compare.py]: This could be an issue since the `input` encoding won't necessarily correspond to the `generated_encoding`. E.g. if `TLDR: ` gets encoded as two token `TLDR:` and ` ` in the `input` but then `TLDR: hello` in `generated_encodings` also gets encoded as two tokens `TLDR:` and ` hello` then you have an issue since you'd be making ` hello` a padding token. 

If you check and this isn't the case, that's fine. You could also just skip the detokenization, tokenization and then it isn't an issue

File: examples/summarization/inference_compare.py [hunk 3]
```diff
@@ -187,39 +188,42 @@ def reward_model(accelerator, model, inputs):
...
 
             for gen_text_even, gen_text_odd, label in zip(generated_texts[::2], generated_texts[1::2], pseudolabels):
-# TODO make TLDR: part of the prompt
-# make sure chosen and rejected don't start with a space
-                prompt = gen_text_even.split("\nTL;DR:")[0]
-                gen_text_even = gen_text_even.split("\nTL;DR:")[1]
-                gen_text_odd = gen_text_odd.split("\nTL;DR:")[1]
+                prompt = gen_text_even.split("\nTL;DR:")[0] + "\nTL;DR:"
+                gen_text_even = gen_text_even.split("\nTL;DR:")[1].strip()
+                gen_text_odd = gen_text_odd.split("\nTL;DR:")[1].strip()
```
Comment (mnoukhov) [inline examples/summarization/inference_compare.py]: If you have the prompt length from the label, I would use that instead of splitting on `TLDR:` since a model could output `TLDR:` multiple times in the output which would create multiple splits. Or set `maxsplit=1`

**Review by mnoukhov — COMMENTED** [review]
Some comments to check out

File: examples/summarization/inference_compare.py (hunk 1 above)
Comment (arianhosseini) [reply examples/summarization/inference_compare.py]: There's already a space there in the `prompt`. I assumed that's the correct format. Right?

File: examples/summarization/inference_compare.py (hunk 2 above)
Comment (arianhosseini) [reply examples/summarization/inference_compare.py]: `prompt_len - 1` should do the trick. input prompt will be something like `[... , 4976,    27, 209]` and generations will be `[4976,    27, 50276, 30592, ....]`.

File: examples/summarization/inference_compare.py (hunk 3 above)
Comment (arianhosseini) [reply examples/summarization/inference_compare.py]: added `maxsplit=1`.

File: examples/summarization/inference_compare.py (hunk 1 above)
Comment (mnoukhov) [reply examples/summarization/inference_compare.py]: The issue is that generation is much better if you don't start with a space (it seems like its an issue in Pythia). So you want to generate without a space `TLDR:` but then when you combine the prompt and completion, it's better to add the space so `TLDR:` + ` ` + `generation`

Comment (mnoukhov) [reply examples/summarization/inference_compare.py]: You should double-check but just follow the format I used in https://huggingface.co/datasets/mnoukhov/openai_summarize_comparisons_relabel_pythia1b_tldrprompt?row=0 There might be no space at all

File: examples/summarization/inference_compare.py (hunk 2 above)
Comment (mnoukhov) [reply examples/summarization/inference_compare.py]: makes sense, you can compute log ps for prompt as well

**Comment by mnoukhov** [comment]
The `inference_generate` script seems to have issues with choosing the right example. After running the model afterwards, I get 78% accuracy (where it should be 100% since we're pseudolabelling with it). Also multi-gpu doesn't seem to work, so I'm just generating different splits separately
